graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -333,3 +343,373 @@ def _process_generic_type_definition(generic_type, schema_id, existing_scalars,
+            source_type_node = type_name_to_definition[outbound_edge_source_type_name]
+            _add_edge_field(
+                source_type_node, outbound_edge_sink_type_name,
+                outbound_field_reference.field_name, inbound_field_reference.field_name,
+                edge_name, OUTBOUND_EDGE_DIRECTION
+            )
+
+        if not cross_schema_edge.out_edge_only:
+            inbound_edge_source_type_names = subclass_sets[inbound_field_reference.type_name]
+            for inbound_edge_source_type_name in inbound_edge_source_type_names:
+                source_type_node = type_name_to_definition[inbound_edge_source_type_name]
+                _add_edge_field(
+                    source_type_node, inbound_edge_sink_type_name,
+                    inbound_field_reference.field_name, outbound_field_reference.field_name,
+                    edge_name, INBOUND_EDGE_DIRECTION
+                )
+
+
+def _check_cross_schema_edge_is_valid(type_name_to_definition, type_name_to_schema_id, scalars,
+                                      union_type_names, cross_schema_edge):
+    """Check that the edge crosses schemas and has valid field references of correct types.
+
+    Args:
+        type_name_to_definition: Dict[str, (Interface/Object)TypeDefinition], mapping
+                                 name of types to their definitions
+        type_name_to_schema_id: Dict[str, str], mapping type name to the id of the schema that the
+                                type is from
+        scalars: Set[str], names of all scalars in the merged_schema, including both built in
+                 and user defined scalars
+        union_type_names: Set[str], names of all union types in the merged schema, used for
+                          informative error messages
+        cross_schema_edge: CrossSchemaEdgeDescriptor namedtuple, the edge that we check the
+                           validity of
+
+    Raises:
+        - InvalidCrossSchemaEdgeError if the cross schema edge lies within one schema, refers
+          to nonexistent schemas, types, or fields, refers to Union types, stitches together
+          fields that are not of a scalar type, or stitches together fields that are of
+          different scalar types
+    """
+    outbound_field_reference = cross_schema_edge.outbound_field_reference
+    inbound_field_reference = cross_schema_edge.inbound_field_reference
+
+    _check_field_reference_is_valid(
+        type_name_to_definition, type_name_to_schema_id, union_type_names, outbound_field_reference
+    )
+    _check_field_reference_is_valid(
+        type_name_to_definition, type_name_to_schema_id, union_type_names, inbound_field_reference
+    )
+
+    if outbound_field_reference.schema_id == inbound_field_reference.schema_id:  # not cross schema
+        raise InvalidCrossSchemaEdgeError(
+            u'Edge "{}" does not cross schemas.'.format(cross_schema_edge)
```

Comment:
nit: Let's tell the user which `schema_id` and which type in that schema it was defined on, so they don't have to look through thousands of lines of schema? Also, we should tell the user why not crossing schemas is a problem -- because we're trying to define cross-schema edges here.

Reply:
`cross_schema_edge` will be printed out with all relevant information (schema_id of both sides, types and fields of both sides) since it's a namedtuple with nice builtin representation, so I don't think it's necessary to additionally repeat this information. I'll add in explanations for why not crossing schemas is bad.